fe/fe-core/src/main/java/com/starrocks/http/HttpServerHandler.java

```diff
@@ -99,28 +100,59 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                 if (LOG.isDebugEnabled()) {
                     LOG.debug("action: {} ", action.getClass().getName());
                 }
-
-                HttpServerHandlerMetrics metrics = HttpServerHandlerMetrics.getInstance();
-                long startTime = System.currentTimeMillis();
-                try {
-                    metrics.handlingRequestsNum.increase(1L);
-                    action.handleRequest(req);
-                } finally {
-                    long latency = System.currentTimeMillis() - startTime;
-                    metrics.handlingRequestsNum.increase(-1L);
-                    metrics.requestHandleLatencyMs.update(latency);
-                    if (latency >= Config.http_slow_request_threshold_ms) {
-                        LOG.warn("receive slow http request. uri: {}, thread id: {}, startTime: {}, latency: {} ms",
-                                WebUtils.sanitizeHttpReqUri(req.getRequest().uri()), Thread.currentThread().getId(),
-                                startTime, latency);
-                    }
+                if (action.supportAsyncHandler()) {
+                    handleActionAsync(req);
+                } else {
+                    handleActionSync(req);
                 }
             }
         } finally {
             ReferenceCountUtil.release(msg);
         }
     }
 
+    private void handleActionSync(BaseRequest request) {
+        RequestWatch watch = new RequestWatch(request);
+        try {
+            action.handleRequest(request);
+        } catch (Exception e) {
+            handleException(request, e);
+        } finally {
+            watch.finish();
+        }
+    }
+
+    private void handleActionAsync(BaseRequest request) {
+        RequestWatch watch = new RequestWatch(request);
```

Comment:
implement interface Closable, and put the RequestWatch in `try (RequestWatch watch = new RequestWatch(request)) {...}`

Reply:
It's not exactly the same as Closable; it needs to be closed within the async phase.